### ISSUE TITLE
BundleDepth set to 1 results in different results than when set to 2

### ISSUE THREAD

**Issue by RALifeCoach** [stealjs/steal-tools#523]
If I set bundleDepth to 2 and run the built files
![bundledepth2](https://cloud.githubusercontent.com/assets/3857096/18189263/40899d68-7087-11e6-8daa-c1b540f305b9.png)
 from https://github.com/RALifeCoach/canjsbuildit is:

If I set bundleDepth to 1 I get the correct number of files built, but when I run it, more files are being read at run time and the results are:
![bundledepth1](https://cloud.githubusercontent.com/assets/3857096/18189225/f085cb3e-7086-11e6-9c45-eb5b3b7c2cb5.png)

**Comment by RALifeCoach** [comment]
What is the status of this item, please?

**Comment by matthewp** [comment]
It's on my todo, it's an important bug, I haven't forgotten it :) Sorry for the delay, will try to get to it as soon as possible.

**Comment by frank-dspeed** [comment]
@matthewp  
I think this can be closed it turned out that package dependency mix up was the problem

i think @RALifeCoach  will aggree

**Comment by matthewp** [comment]
Oh wow, can you summarize what fixed it?

**Comment by frank-dspeed** [comment]
Deleting Node Modules fresh re install did fix it.

**Comment by RALifeCoach** [comment]
No it is not fixed. Please investigate.

**Comment by frank-dspeed** [comment]
I Still think its solved because:

The Bundel Results in:
app,js - containing bundel information all canjs
main1.js - the individual js code for the main1 component
main2.js - the individual js code for the main2 component

both gets loaded as it should and have correct size 

https://github.com/frank-dspeed/canjsbuildit/tree/master/dist/bundles

watch the result it has nothing doubled and is 100% use able for his use case.

he whants to create a CMS where he puts in components on the same page and then only needs to load 1 js file per page and 1 addition for each component.

at present we created 3 files and a file for each compoent.

2 files with the diffrent sets over the app main1-main2 and main1-main2-main3 they are realy small
then we have bundlit/app.js with the rest of the code and one file for each component

so he is complaining about 2 js and 2 css files! that are to much 

i don't think thats redesign able or should even be redisigned.

**Comment by pYr0x** [comment]
talked to @frank-dspeed. this issue is solved. OP was using an old steal version.
pls open an new issue if an other issue is still present

**Comment by phillipskevin** [comment]
re-opening until this is resolved for @RALifeCoach

**Comment by RALifeCoach** [comment]
I will consider this closed when steal-tools builds exactly 4 JS files (app.js, main1.js, main2.js, main3.js) and when I run the built files it produces the correct output as shown in the initial image in the original post.

**Comment by matthewp** [comment]
Looking into this now.

**Comment by matthewp** [comment]
Added a test in 4d6f59fa89327c0ca4b3a032d5c8f9903988b88c that shows the issue, apparently shared modules between bundles are not being added to any bundle when bundlesDepth: 1 is used.

**Comment by matthewp** [comment]
Since we are not allowing any shared bundles steal-tools doesn't know where to put those shared dependencies. Investigating.

**Comment by matthewp** [comment]
So, a simpler description of what is going on here. Let's say you have bundles like this:

```js
- a
  - app/a
  - app/c
- b
  - app/b
  - app/c
```

Since you are saying that you want to create, at most, 1 bundle for each bundle, where are shared dependencies supposed to go? In this case `app/c` is a shared dependency between the two bundles. What is currently happening (the bug) is that it's not being included in any bundle; again because it has no where to go.

The fix will be to allow top-level bundles to be merged together, so that a and b because a-b which contains `app/a`, `app/b` and `app/c`.

**Comment by matthewp** [comment]
Let's lay this out again and describe what I see as 3 possible solutions:

```js
- a
  - app/a
  - app/c
- b
  - app/b
  - app/c
```

A. Bundles a and b are merged together as `a-b`.  Estimate: *1* (already solved)
B. `app/c` is put into either `a` or `b`.  Estimate *3*
C. `app/c` is put into the main bundle.  Estimate *6*

**Comment by frank-dspeed** [comment]
@matthewp what you don't understand and i don't understood also is that his goal is not to use steal anymore after that so its a export he whants a method that creates 4 files 1 main file and then 1 file with all dependencys for each component like one is using jquery and all that he whants a main bundle and then component bundles 

but as a export!!! after that he will load that via script tags main.js and add later in code on some places in his cms script component.js and expects then all rendered right.

**Comment by pYr0x** [comment]
@matthewp i prefer solvitions **A**. i think this solutions is the best way to share deps over different modules.
Solution **C** could maybe also good. But the question is, is it worth?

**Comment by matthewp** [comment]
I have a fix that will use option `A` (listed in https://github.com/stealjs/steal-tools/issues/523#issuecomment-258431863) as it's the easiest and cleanest fix at the moment.  Will submit a PR for that shortly.

@RALifeCoach I'm curious about what you're trying to solve by lowering the bundleDepth to 1. Are you trying to prevent the waterfall that occurs?  If so, you might want to check out #541, which lays out an idea that will allow you to list all of your bundles in script tags, so the requests can be parallelized, and there would be no waterfall. Would this be something that would solve your problem?

**Comment by RALifeCoach** [comment]
What I want is for the main to hold all the CanJS code and any other code that is common to ALL modules. Then I want each module to load in a single file. I.e. if I have a page that includes 3 CanJS based components then the page would load exactly 4 js files (app.js, comp1.js, comp2.js, comp3.js). And then I would like the code to work when I run it. It doesn't matter to me if steal is included or not.

**Comment by matthewp** [comment]
@RALifeCoach Yeah, the bug is about what happens with `app.js` and `comp1.js` share 1 dependency. Like you said, you only want to load at max 4 files, so where do those shared dependencies go? They can't go into the main, they are not common to all of the bundles, only 2 of the bundles.

The solution I'm going to release will merge together the `app` and `comp1` bundles. So instead of 4 additional JS requests, there will be only 3.

**Comment by RALifeCoach** [comment]
If there are items that are common to comp1 and comp2, but not comp3 then they would be in comp1 AND in comp2.

**Comment by pYr0x** [comment]
but that is not the philosophy of shared dependencies...

**Comment by RALifeCoach** [comment]
@pYr0x if there were 50 modules with many interdependencies and the bundleDepth was set to 3, what do you do with the rare, shared dependencies? At some point in time, the sharing stops. With bundleDepth 1 it just stops sooner.

**Comment by matthewp** [comment]
The sharing never stops, it just moves where the sharing happens. If you had bundles like:

```
a
a-b
a-b-c
a-b-c-d
a-b-c-d-e
a-b-c-d-e-f
```

With bundleDepth of 3, the last 3 bundles' dependencies just get moved up to the parent bundles. this is only an issue because of bundleDepth 1 where there's no good place to put shared dependencies; they have to go *somewhere*.

I did think about duplication as a solution. My biggest concerns there (other that it would be a bigger change, steal-tools never duplicates currently), is I don't know how AMD handles a module that registers itself more than once.

**bundles/a**
```js
define("app/a", function() { ... })
define("app/c", function() { ... })
```

**bundles/b**
```js
define("app/a", function() { ... })
define("app/c", function() { ... })
```

What happens when the 2nd app/c loads here? I don't know if AMD complains, or if it overwrites the 1st app/c or if it ignores the 2nd one... we would want it to ignore the 2nd, but I don't know for sure if that's what it does or not.

**Comment by RALifeCoach** [comment]
Maybe it would be better to issue an error is bundleDepth is < 2. For my purpose, I will export a general purpose library and then use 'can' from the global namespace - just like I would have using 2.3.

**Comment by matthewp** [comment]
Nah, we should cover it, it's just deciding what is correct for 1 and 0 that is hard. For 0 it means create no bundles at all (everything should go in the main).

**Comment by matthewp** [comment]
Closed by https://github.com/stealjs/steal-tools/pull/542 and released in `steal-tools@1.0.0-rc.4`
